### ISSUE TITLE
New feature suggestion

### ISSUE THREAD

**Issue by laminsk** [Bilibili/ijkplayer#2375]
Though I'm not an experienced ffmpeg developer, I've been following ijkplayer for sometimes for my pet project and also for learning. I do think this project should support caching or even some way to save the stream like Exoplayer does. I understand that IMediasource is available for this purpose but it's not straight forward compared to Exoplayer. Just a suggestion really. Cheers.

**Comment by xinzhengzhang** [comment]
The [cache](https://www.ffmpeg.org/ffmpeg-protocols.html#cache) seems what you need. But for now the cache protocol is not support in mobile device(Android, iOS). Because `/tmp/xxx` is not allowed to write for permission issues. 
Thanks for your suggestion and I will make a patch to ffmpeg for supporting passing a specified directory.

**Comment by laminsk** [comment]
Hi, thanks for the reply. I think it'd be useful to write the decoded stream to local device when playing media files from web. For instance;

1) pass the video url (hosted on webserver) to ijkplayer 
2) ijkplayer starts the playback after buffering enough data
3) while doing so, write the decoded stream data to output file in device storage (we specify the location).

**Comment by xinzhengzhang** [comment]
What you said has been implemented in `cache` in ffmpeg except specifying the storage location and I will improve it (may be next release) to ffmpeg.

HOW TO DO
1. add `cache` prefix in your video video url.  `http://xxx.yyy.zzz/1.flv`  => `cache:http://xxx.yyy.zzz/1.flv` 
2. pass transferred url to ijkplayer
- - -
3. using option to specify the storage location to ijkplayer (NOT IMPLEMENTED YET)

**Comment by laminsk** [comment]
Hi there

Thanks for the follow up. What you suggested sounds good. May I ask whether it's straight forward to access the cached data? For instance, I understand that by using "cache:http://xxx.yyy.zzz/1.flv", data will be cached on the device, but how can we actually retrieve that data for our own use? 

Using cachedatasource or okhttpdatasource allows to access the data from source just like IMediasource is designed to do. What I'm trying to achieve is:

1) pass the video url (hosted on webserver) to ijkplayer (http://xxx.yyy.zzz/1.flv)
2) ijkplayer starts the playback after buffering enough data from the url. At the same time, ijkplayer will write the streamed data (inputstream) to local storage. (we specify the location and file name).
3) or perhaps after specifying ijkplayer.datasource("url"), we could use something like  ijkplayer.copy(inputstream). We can then do whatever we want it after that.

**Comment by xinzhengzhang** [comment]
The cache protocol does not save media data continually,
I don't think it would work, if you need to access the media data.

A custom implement of `IMediasource` is the best choice for you, for now.

**Comment by zldada** [comment]
@xinzhengzhang @laminsk It seems like that all videodata which read from local file or network will be read by io_read_packet() defined in /ffmpeg/libavformat/aviobuf.c. So,while the video is playing,I try to save the videodata into sdcard.It is ok for me to play the saved video which read from local video file,but I failed to play the saved video which read from network. I have compared the original video to the saved video,found that the size of saved video which read from network is always bigger than the original,but the local one does't !.I don't know how to solve this problem.
 I wonder that this may be a possible way to save streaming data.

**Comment by xinzhengzhang** [comment]
The streaming data is not continually

**Comment by zldada** [comment]
@xinzhengzhang you are right. I have corrected my code,and now I can save the streaming data and play the saved video successfully.

**Comment by laminsk** [comment]
@zldada I'd appreciate if you could provide an example code regarding saving data from network to sdcard. Thanks so much in advance.

**Comment by zldada** [comment]
@laminsk All you have to do is that add some code used to save the buffer into sdcard. just modify   io_read_packet() defined in /ffmpeg/libavformat/aviobuf.c and recompile ffmpeg and ijkplayer

**Comment by laminsk** [comment]
@zldada Thanks for the reply. I'm not an experienced C coder and I simply trying to figure this out for my pet project. Any pointers to get this working will be greatly appreciated.

**Comment by codeliu6572** [comment]
Thank you，I see.

**Comment by waddwaw** [comment]
@zldada Can you can record video projects under the share?

**Comment by OCer** [comment]
加了cache:无法播放  用的0.7.7版本  编译的是支持最多格式  url cache:rtmp://live.hkstv.hk.lxdns.com/live/hks  @xinzhengzhang
